### ISSUE TITLE
whitening

### ISSUE THREAD

**Issue by RuibinMa** [filipradenovic/cnnimageretrieval-pytorch#8]
Dear Filip,
I'm wondering how the "custom computed" initialization weights for the whitening layers:
whiten = nn.Linear(dim, dim, bias=True)
are computed, such as those in here: http://cmp.felk.cvut.cz/cnnimageretrieval/data/whiten/retrieval-SfM-120k/
Is there a tutorial for that? I'm trying to create my own dataset.
Thanks in advance!

**Comment by filipradenovic** [comment]
You can find an example of how to learn the whitening in our test script [#L124](https://github.com/filipradenovic/cnnimageretrieval-pytorch/blob/c4fca8958631c03cefff9e8ae6bfad4d9813b633/cirtorch/examples/test.py#L124). Then, using computed `m` and `P` your whitening Linear layer is initialized as:
```
Lw = {'m': torch.from_numpy(m).float(), 'P': torch.from_numpy(P).float()}
dim = net.meta['outputdim'] #OUTPUT_DIM[model]
whiten = nn.Linear(dim, dim, bias=True)
whiten.weight.data = Lw['P']
whiten.bias.data = -torch.mm(Lw['P'], Lw['m']).squeeze()
```

**Comment by RuibinMa** [comment]
Hi Filip,
Thanks for the reply. I noticed that in test, you learn a whitening based on current network's output (https://github.com/filipradenovic/cnnimageretrieval-pytorch/blob/c4fca8958631c03cefff9e8ae6bfad4d9813b633/cirtorch/examples/test.py#L151). But upon initializing the network, you used precomputed whitening weights. This looks to me a chicken-and-egg problem. How should I get the whitening weights to initialized networks when I start from scratch?

**Comment by filipradenovic** [comment]
We start training from the off-the-shelf network (weights learned on imagenet) so you would learn whitening using output of the off-the-shelf network to initialize it. This is how the before-mentioned precomputed whitening were computed. If you want to train network from scratch (random weight initialization) then whitening should be randomly initialized as well.

**Comment by RuibinMa** [comment]
I see. Thanks a lot!
